### PR TITLE
spring-boot-globalExceptionHandler-and-data-validation-solution1

### PR DESCRIPTION
**Pull request by DojiDoe**

### CODE REVIEW

File: src/main/java/mate/academy/bookstore/validation/Isbn.java [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+import java.lang.annotation.Target;
+
+@Constraint(validatedBy = IsbnValidator.class)
+@Target({ElementType.PARAMETER, ElementType.FIELD})
+@Retention(RetentionPolicy.RUNTIME)
+public @interface Isbn {
+    String message() default "Invalid format isbn";
+    Class<?>[] groups() default {};
+    Class<? extends Payload>[] payload() default {};
+}
```
Comment (ahoienko) [inline src/main/java/mate/academy/bookstore/validation/Isbn.java]: Please check is annotation that validates ISBN already exists, to not implement logic that already implemented

File: src/main/java/mate/academy/bookstore/dto/CreateBookRequestDto.java [hunk 2]
```diff
@@ -1,16 +1,25 @@
...
 import lombok.Setter;
+import mate.academy.bookstore.validation.Isbn;
 
 @Getter
 @Setter
 public class CreateBookRequestDto {
+    @NotNull
     private String title;
+    @NotNull
     private String author;
```
Comment (ahoienko) [inline src/main/java/mate/academy/bookstore/dto/CreateBookRequestDto.java]: Is it ok if string fields of this model will be blank "" or "    "?

**Review by ahoienko — CHANGES_REQUESTED** [review]
Very nice) Lets improve your solution

File: src/main/java/mate/academy/bookstore/controller/BookController.java [hunk 3]
```diff
@@ -44,7 +45,7 @@ public void delete(@PathVariable Long id) {
     }
 
     @PostMapping
-    public BookDto createBook(@RequestBody CreateBookRequestDto book) {
+    public BookDto createBook(@RequestBody @Valid CreateBookRequestDto book) {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/controller/BookController.java:49]: Didn't find update book endpoint, please add it

File: src/main/java/mate/academy/bookstore/dto/CreateBookRequestDto.java [hunk 4]
```diff
@@ -1,16 +1,30 @@
...
+import jakarta.validation.constraints.Size;
 import java.math.BigDecimal;
 import lombok.Getter;
 import lombok.Setter;
+import org.hibernate.validator.constraints.ISBN;
 
 @Getter
 @Setter
 public class CreateBookRequestDto {
+    @NotNull
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/dto/CreateBookRequestDto.java]:
```suggestion
    @NotBlank
```

for all String

File: src/main/java/mate/academy/bookstore/dto/CreateBookRequestDto.java [hunk 5]
```diff
@@ -1,16 +1,30 @@
...
+    @Pattern(regexp = "^(?=\\s*\\S).*$")
     private String title;
+    @NotNull
+    @Pattern(regexp = "^(?=\\s*\\S).*$")
     private String author;
+    @ISBN
     private String isbn;
+    @Min(0)
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/dto/CreateBookRequestDto.java:23]:
```suggestion
    @Min(0)
    @NotNull
```

File: src/main/java/mate/academy/bookstore/exception/CustomGlobalExceptionHandler.java [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+            WebRequest request
+    ) {
+        Map<String, Object> body = new LinkedHashMap<>();
+        body.put("timestamp", LocalDateTime.now());
+        body.put("status", HttpStatus.BAD_REQUEST);
+        List<String> errors = ex.getBindingResult().getAllErrors().stream()
+                .map(this::getErrorMessage)
+                .toList();
+        body.put("errors", errors);
+        return new ResponseEntity<>(body, headers, status);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/exception/CustomGlobalExceptionHandler.java]:
```suggestion
        return new ResponseEntity<>(body, headers, HttpStatus.BAD_REQUEST);
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/mate/academy/bookstore/dto/CreateBookRequestDto.java [hunk 7]
```diff
@@ -1,16 +1,30 @@
...
 @Getter
 @Setter
 public class CreateBookRequestDto {
+    @NotNull
+    @Pattern(regexp = "^(?=\\s*\\S).*$")
     private String title;
+    @NotNull
+    @Pattern(regexp = "^(?=\\s*\\S).*$")
     private String author;
+    @ISBN
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/dto/CreateBookRequestDto.java:21]: check if `@ISBN` includes NotNull checking, if not add it

File: src/main/java/mate/academy/bookstore/controller/BookController.java (hunk 3 above)
Comment (Elena-Bruyako) [reply src/main/java/mate/academy/bookstore/controller/BookController.java:49]: comment not fixed

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]

File: src/main/java/mate/academy/bookstore/controller/BookController.java (hunk 3 above)
Comment (DojiDoe) [reply src/main/java/mate/academy/bookstore/controller/BookController.java:49]: image

Comment (DojiDoe) [reply src/main/java/mate/academy/bookstore/controller/BookController.java:49]: There is a Valid annotation

Comment (Elena-Bruyako) [reply src/main/java/mate/academy/bookstore/controller/BookController.java:49]: I didn't text about Valid annotation anything. 
BookController doesn't have update endpoint, so please add it

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]

**Review by sarakhmen — APPROVED** [review]

**Pull request merged**